File: src/+estimation/iterative_estimator.m
classdef iterative_estimator < handle
    properties
        sensor_list
        last_location
        d_s
        Dd_s
    end
    methods
        function obj = iterative_estimator(sensor_list, initial_location_guess)
            arguments
                sensor_list
                initial_location_guess
            end
            obj.sensor_list = sensor_list;
            obj.last_location = initial_location_guess;
            
            % Create x variable to differentiate
            x = symmatrix('x', size(initial_location_guess));
            
            
            % Create symbolic sensor location to let us differentiate once and not for every sensor
            sens_loc = symmatrix('sens_loc', size(initial_location_guess));
            % Symbolic function to differentiate
            d = ((x-sens_loc) * (x-sens_loc).').^0.5;
            % Create symbolic functions that can get parameters (matrix symbolic functions can't use parameters)
            
            d_s(symmatrix2sym(sens_loc), symmatrix2sym(x)) = symmatrix2sym(d);
            Dd = diff(d,x);
            Dd_s(symmatrix2sym(sens_loc), symmatrix2sym(x)) = symmatrix2sym(Dd);
            obj.d_s = matlabFunction(d_s);
            obj.Dd_s = matlabFunction(Dd_s);
        end

        function [dist, Ddist] = get_distances(obj, sensor_locations, x0)
            % when calculating by hand:
            % distances = ((x0(1) - sensor_locations(:,1)).^2 + (x0(2) - sensor_locations(:,2)).^2 + (x0(3) - sensor_locations(:,3)).^2).^0.5;
            % d_distances = (x0 - sensor_locations) ./ distances;

            % Convert row to column vectors
            sensor_locations = sensor_locations';
            % Example of how vectors should look like:
            % sensor_locations = [0,0,0; 0,1,0; 1,0,0; -1,-1,0]';
            % x0 = [1, 1, 0];

            x0_c = num2cell(x0);

            % Create location to put output vector of symbols
            f = zeros([1 size(sensor_locations, 2)]);
            Df = zeros([size(x0,2) size(sensor_locations, 2)]);
            
            % Calculate function and derivative for each sensor
            for i = 1:length(f)
                sens_loc_c = num2cell(sensor_locations(:,i));
                f(i) = obj.d_s(sens_loc_c{:}, x0_c{:});
                Df(:,i) = obj.Dd_s(sens_loc_c{:}, x0_c{:});
            end
            % Convert vector of symbols to numbers
            dist = f';
            Ddist = Df';
        end

        function point = estimate_point_by_distances(obj, distances, sensor_locations, x0)
            current_estimate = x0';
            last_estimate = current_estimate;
            y = distances;
            for i = 1:5
                [h_x0, H] = obj.get_distances(sensor_locations, current_estimate');
                
                % To calculate how close the H matrix is to being singular, uncomment this section
                % [s, v, d] = svd(H'*H);
                % a = diag(v);
                % kappa = a(1)/a(end);
                % display(kappa);
                
                % Add a small value to diagonal of H to prevent it from being singular
                H = H + eye(size(H))*0.1;

                % Make sure the matrix we use can be inverted
                % (this is probably wasteful as it's probably quicker to just invert a 3x3 matrix in matlab than to test an if statement)
                % if (size(H,1) == size(H,2))
                %     Z = inv(H);
                % else
                %     Z = inv(H'*H)*H';
                % end
                % (the above is probably wasteful as it's probably quicker to just multiply a 3x3 matrix twice in matlab than to test an
                % if statement, so let's just multiply)
                Z = inv(H'*H)*H';
                
                % Get next estimate
                current_estimate = Z*(y - h_x0 + H*current_estimate);
                
                % temp_estimate = inv(H + eye(size(H))*0.1)*(y-h_x0) + current_estimate;
                % display(current_estimate);
                % display(temp_estimate);
                % display(current_estimate);

                % current_estimate = temp_estimate;
                if (norm(last_estimate - current_estimate) < 0.01)
                    break;
                end
                last_estimate = current_estimate;

                
            end
            % loss = (h_x0 - distances)';
            % display(loss);
            point = current_estimate';
        end

        function estimated_path = estimate_path_by_distance(obj)
            arguments
                obj
            end

            all_distances = zeros([size(obj.sensor_list, 2) size(obj.sensor_list(1).noisy_distances, 1)]);
            sensor_locations = zeros([size(obj.sensor_list, 2) size(obj.sensor_list(1).sensor_position, 2)]);
            estimated_path = zeros([size(obj.sensor_list(1).noisy_distances, 1) size(obj.sensor_list(1).sensor_position, 2)]);
            for i = 1:size(obj.sensor_list, 2)
                all_distances(i,:) = obj.sensor_list(i).noisy_distances;
                sensor_locations(i,:) = obj.sensor_list(i).sensor_position;
            end
            % estimated_path = obj.estimate_point_by_distances(all_distances(:,1), sensor_locations, obj.last_location);
            f = waitbar(0, "please wait");
            for i = 1:size(all_distances, 2)
                current_point = obj.estimate_point_by_distances(all_distances(:,i), sensor_locations, obj.last_location);
                obj.last_location = current_point;
                estimated_path(i,:) = current_point;
                waitbar(i/size(all_distances, 2), f, "processing");
            end
            close(f);
        end
    end
end

File: python/src/estimation/iterative_estimator.py

import numpy as np

class iterative_estimator:
    def __init__(self, sensors, initial_guess_position):

        # Initialize parameters
        self.sensor_locations = sensors.sensor_locations
        self.distances = sensors.noisy_distances
        self.last_location = np.array(initial_guess_position)

        # Initialize functions
        self.dist_func = lambda sensor_pos, pos: (np.linalg.norm(pos - sensor_pos, 2, 1))
        self.dist_derivative_func = lambda sensor_pos, pos: ((pos - sensor_pos).T / np.linalg.norm(pos - sensor_pos, 2, 1)).T

    def estimate_point_by_distance(self, distances, last_estimate, iteration_limit=5, precision_limit=0.01, epsilon=0.1):
        # Initialize parameters with correct dimensions
        last_estimate = last_estimate.T
        current_estimate = last_estimate
        y = distances.T


        # Iterate until we get close enough to the point
        for i in range(iteration_limit):
            # Get distance and derivative for current point
            h_x0 = self.dist_func(self.sensor_locations, current_estimate.T)
            H = self.dist_derivative_func(self.sensor_locations, current_estimate.T)

            # Make sure H isn't singular
            H = H + np.eye(H.shape[0],H.shape[1]) * epsilon
            Z = np.linalg.inv(H.T @ H) @ H.T
            # Get better estimate
            current_estimate = Z @ (y - h_x0 + H @ current_estimate)

            # stop if we don't get closer
            if np.linalg.norm(last_estimate - current_estimate) < precision_limit:
                break
            last_estimate = current_estimate
        
        return current_estimate.T
    
    def estimate_path(self, override_path_init=None, iteration_limit=5, precision_limit=0.01):
        estimated_path = np.zeros([self.distances.shape[0], self.sensor_locations.shape[1]])
        for i in range(self.distances.shape[0]):
            if override_path_init is not None:
                self.last_location = override_path_init[i,:]
            current_estimate = self.estimate_point_by_distance(self.distances[i,:], self.last_location, iteration_limit=iteration_limit, precision_limit=precision_limit)
            self.last_location = current_estimate
            estimated_path[i,:] = current_estimate
        return estimated_path
    
    def get_cov_err(self, true_path):
        path_cov_err = np.zeros(true_path.shape)
        for i in range(true_path.shape[0]):
            current_point = true_path[i,:]
            H = self.dist_derivative_func(self.sensor_locations, current_point)
            cov_vec = np.diag(np.linalg.inv(H.T @ H))
            path_cov_err[i,:] = cov_vec.T
        return path_cov_err
